theano/compile/builders.py

```diff
@@ -137,6 +137,46 @@ def perform(self, node, inputs, outputs):
             # we wont need this copy anymore
             output[0] = variable.copy()
 
+    def c_code(self, node, name, inputs, outputs, sub):
+        order = io_toposort(self.new_inputs, self.new_outputs)
+
+        # assert all inner nodes have c_code()
+        if not (self._op_use_c_code and
+                all([getattr(node.op, '_op_use_c_code', False) and
+                     hasattr(node.op, 'c_code') for node in order])):
+            self._op_use_c_code = False
+            return
+        else:
+            c_code = ""
+            local = locals()
+            for inner_node in order:
+                name = 'node' + str(id(inner_node))
+
+                in_names = []
+                for i, var in enumerate(inner_node.inputs):
+                    if var in self.new_inputs:
+                        in_names.append(inputs[i])
+                    else:
+                        new_name = 'InVar' + str(id(var))
+                        in_names.append(new_name)
+                        local.setdefault(new_name, var.clone())
+                        c_code += var.type.c_declare(new_name, sub)
+
+                out_names = []
+                for i, var in enumerate(inner_node.outputs):
+                    if var in self.new_outputs:
+                        out_names.append(outputs[i])
+                    else:
+                        new_name = 'OutVar' + str(id(var))
+                        out_names.append(new_name)
+                        local.setdefault(new_name, var.clone())
+                        c_code += var.type.c_declare(new_name, sub)
+
+                c_code += inner_node.op.c_code(inner_node, name,
```

Comment:
I don't think this is the good strategy. I should have gave you more guideline. Sorry for this.

There is already a linker that compile a function into a full thunk. Mode(linker='c') generate that. This use the CLinker class. I think it is better to not try to duplicate all that code.

I'm not sure how to reuse it correctly. @abergeron you started work on that for scan I think. Do you still have that around?